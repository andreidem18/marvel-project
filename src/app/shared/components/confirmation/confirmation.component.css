.confirmation {
    min-width: 300px;
    max-width: 500px;
}

.message-container {
    padding: 30px;
    text-align: center;
}

.buttons-container {
    display: flex;
    height: 60px;
}

.buttons-container button {
    width: 50%;
    font-size: 1.1rem;
    font-weight: 700;
    /* color: var(--primary); */
    background: var(--surface-dark);
}

.buttons-container button:hover {
    filter: brightness(103%);
    transition: .1s;
}

.buttons-container button.no {
    color: var(--font-color-contrast);
    background: var(--secondary);
}
