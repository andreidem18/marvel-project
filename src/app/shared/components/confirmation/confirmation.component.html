<shared-modal-wrapper [isOpen]="isOpen" (isOpenChange)="close()">
    <div class="confirmation">
        <div class="message-container">
            <div [innerHTML]="confirmation.message"></div>
        </div>
        <div class="buttons-container">
            <button class="no" (click)="close()">no</button>
            <button class="yes" (click)="fireAction()">yes</button>
        </div>
    </div>
</shared-modal-wrapper>
