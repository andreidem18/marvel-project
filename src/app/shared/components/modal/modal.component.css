.modal-container {
    position: fixed;
    visibility: hidden;
    opacity: 0;
    transition: visibility .2s, opacity .2s linear;
    z-index: 9;
}

.modal-container.active {
    visibility: visible;
    opacity: 1;
    transition: visibility .2s, opacity .2s linear;
}

.modal-container .overlay {
    position: absolute;
    background: rgba(0,0,0,.3);
}


.modal-container .modal-body {
    position: relative;
    z-index: 9;
    background: var(--surface-light);
}
