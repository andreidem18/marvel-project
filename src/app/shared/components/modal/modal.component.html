<div 
    class="modal-container wrap center"
    [ngClass]="{'active': isOpen}"
>
    <div class="modal-body">

        <ng-content ></ng-content>
        
    </div>
    <div 
        class="overlay wrap"
        (click)="close()"
    ></div>
</div>
