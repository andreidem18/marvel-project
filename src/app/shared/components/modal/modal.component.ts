import { Component, EventEmitter, Input, Output } from '@angular/core';

@Component({
  selector: 'shared-modal-wrapper',
  templateUrl: './modal.component.html',
  styleUrls: ['./modal.component.css']
})
export class ModalComponent {

  @Input()
  public isOpen: boolean = true;

  @Output()
  public isOpenChange = new EventEmitter<boolean>();

  close() {
    this.isOpen = false;
    this.isOpenChange.emit(false);
  }

}
