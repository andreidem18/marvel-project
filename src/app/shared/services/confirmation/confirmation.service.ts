import { Injectable } from '@angular/core';
import { Confirmation } from '../../interfaces/Confirmation';

@Injectable({
  providedIn: 'root'
})
export class ConfirmationService {

  private _isOpen: boolean = false;

  private defaultConfirmation: Confirmation = {
    action: () => {},
    message: ''
  };
  private _confirmationInfo: Confirmation = this.defaultConfirmation;

  get isOpen() { return this._isOpen }
  get confirmationInfo() { return { ...this._confirmationInfo } }

  constructor() { }

  display(confirmation: Confirmation) {
    this._isOpen = true;
    this._confirmationInfo = {
      ...confirmation,
      action: () => {
        confirmation.action();
        this.close();
      }
    };
  }

  close() {
    this._isOpen = false;
    setTimeout(() => {
      this.clearConfirmation();
    }, 800);
  }

  clearConfirmation() {
    this._confirmationInfo = this.defaultConfirmation;
  }
}
