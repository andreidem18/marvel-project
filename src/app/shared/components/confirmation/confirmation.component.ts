import { Component, inject } from '@angular/core';
import { ConfirmationService } from '../../services/confirmation/confirmation.service';
import { Confirmation } from '../../interfaces/Confirmation';

@Component({
  selector: 'shared-confirmation',
  templateUrl: './confirmation.component.html',
  styleUrls: ['./confirmation.component.css']
})
export class ConfirmationComponent {

  private confirmationService = inject(ConfirmationService);

  get confirmation(): Confirmation {
    return this.confirmationService.confirmationInfo;
  }

  get isOpen(): boolean {
    return this.confirmationService.isOpen;
  }

  fireAction() {
    this.confirmation.action();
  }

  close() {
    this.confirmationService.close();
  }

}
