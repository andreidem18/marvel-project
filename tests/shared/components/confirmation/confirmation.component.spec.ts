import { ComponentFixture, TestBed } from '@angular/core/testing';

import { ConfirmationComponent } from '../../../../src/app/shared/components/confirmation/confirmation.component';
import { ConfirmationService } from 'src/app/shared/services/confirmation/confirmation.service';
import { confirmationMock } from '../../mock-data/confirmation-mock';
import { ModalComponent } from 'src/app/shared/components/modal/modal.component';

describe('ConfirmationComponent', () => {
  let component: ConfirmationComponent;
  let fixture: ComponentFixture<ConfirmationComponent>;
  let service: ConfirmationService;
  let compiled: HTMLElement;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [ConfirmationComponent, ModalComponent],
    });
    fixture = TestBed.createComponent(ConfirmationComponent);
    service = TestBed.inject(ConfirmationService);
    compiled = fixture.nativeElement;
    
    service.display(confirmationMock);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match with the snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should trigger the confirmation action when clicking yes button', () => {
    jest.spyOn(component, 'fireAction');

    const buttonTag = compiled.querySelector('.yes');
    expect(buttonTag).toBeTruthy();
    buttonTag!.dispatchEvent(new Event('click'));

    expect(component.fireAction).toHaveBeenCalled();
  });

  it('should close when clicking the no button', () => {
    const buttonTag = compiled.querySelector('.no');
    buttonTag?.dispatchEvent(new Event('click'));
    expect(service.isOpen).toBeFalsy();
  });

  it('should close when clicking the yes button', () => {
    const buttonTag = compiled.querySelector('.yes');
    buttonTag?.dispatchEvent(new Event('click'));
    expect(service.isOpen).toBeFalsy();
  });
});
