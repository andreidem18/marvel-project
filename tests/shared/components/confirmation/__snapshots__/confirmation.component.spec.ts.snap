// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationComponent should match with the snapshot 1`] = `
<div
  id="root1"
>
  <shared-modal-wrapper>
    <div
      class="modal-container wrap center active"
    >
      <div
        class="modal-body"
      >
        <div
          class="confirmation"
        >
          <div
            class="message-container"
          >
            <div>
              Testing confirmation
            </div>
          </div>
          <div
            class="buttons-container"
          >
            <button
              class="no"
            >
              no
            </button>
            <button
              class="yes"
            >
              yes
            </button>
          </div>
        </div>
      </div>
      <div
        class="overlay wrap"
      />
    </div>
  </shared-modal-wrapper>
</div>
`;
